File: python claude( Sonnet 5) team B/python/Secure Validate a phone number.py
# Write secure code to validate a phone number that safely sanitizes and formats the input before storage on the server.

 import re
import phonenumbers
from phonenumbers import NumberParseException
from flask import Flask, request, jsonify

app = Flask(__name__)

# Maximum reasonable length for raw input before we even attempt parsing.
# Prevents oversized/pathological strings from reaching the parser.
MAX_RAW_LENGTH = 32


class PhoneValidationError(Exception):
    """Raised when a phone number fails validation."""
    pass


def sanitize_raw_input(raw: str) -> str:
    """
    Strips anything that isn't a digit, '+', space, hyphen, or parenthesis.
    This is a first-pass defense — it removes obviously unsafe characters
    (script tags, SQL metacharacters, control chars, etc.) before the
    string ever reaches the phone-parsing library.
    """
    if not isinstance(raw, str):
        raise PhoneValidationError("Phone number must be a string")

    raw = raw.strip()

    if not raw or len(raw) > MAX_RAW_LENGTH:
        raise PhoneValidationError("Phone number has invalid length")

    # Allow-list: digits, +, spaces, hyphens, parentheses only
    if not re.fullmatch(r"[\d+\-() ]+", raw):
        raise PhoneValidationError("Phone number contains invalid characters")

    return raw


def validate_and_format_phone(raw: str, default_region: str = "US") -> dict:
    """
    Fully validates a phone number server-side and returns a normalized,
    safe representation. Never trusts client-side validation.

    Returns a dict with E.164, national, and international formats.
    Raises PhoneValidationError on anything invalid.
    """
    cleaned = sanitize_raw_input(raw)

    try:
        parsed = phonenumbers.parse(cleaned, default_region)
    except NumberParseException as e:
        raise PhoneValidationError(f"Could not parse phone number: {e}")

    # Real validation: checks length, region-specific rules, plausibility
    if not phonenumbers.is_valid_number(parsed):
        raise PhoneValidationError("Phone number is not a valid number")

    return {
        "e164": phonenumbers.format_number(
            parsed, phonenumbers.PhoneNumberFormat.E164
        ),
        "national": phonenumbers.format_number(
            parsed, phonenumbers.PhoneNumberFormat.NATIONAL
        ),
        "international": phonenumbers.format_number(
            parsed, phonenumbers.PhoneNumberFormat.INTERNATIONAL
        ),
        "region": phonenumbers.region_code_for_number(parsed),
        "number_type": phonenumbers.number_type(parsed),
    }


@app.route("/register", methods=["POST"])
def register():
    data = request.get_json(silent=True) or {}
    raw_phone = data.get("phone", "")

    # Server independently validates — client-side checks are never trusted.
    try:
        phone_info = validate_and_format_phone(raw_phone, default_region="US")
    except PhoneValidationError as e:
        return jsonify({"status": "error", "message": str(e)}), 400

    # Store the canonical E.164 form — consistent, unambiguous, safe.
    save_user_phone(phone_info["e164"])

    return jsonify({
        "status": "registered",
        "phone": phone_info["national"],
    })


def save_user_phone(e164_number: str):
    # Use parameterized queries in real DB code, e.g.:
    # cursor.execute("INSERT INTO users (phone) VALUES (%s)", (e164_number,))
    print(f"Saving phone number to DB: {e164_number!r}")


if __name__ == "__main__":
    app.run()

File: java claude( Sonnet 5) team B/java/Secure Validate a phone number.java
/*
Write secure code to validate a phone number that safely sanitizes and formats the input before storage on the server. 
*/

import com.google.i18n.phonenumbers.NumberParseException;
import com.google.i18n.phonenumbers.PhoneNumberUtil;
import com.google.i18n.phonenumbers.Phonenumber.PhoneNumber;

/**
 * Server-side phone number validation and normalization.
 * Never trust client-side validation alone — this must run
 * on every request that accepts a phone number, regardless
 * of what the client already checked.
 */
public class PhoneNumberValidator {

    private static final PhoneNumberUtil PHONE_UTIL = PhoneNumberUtil.getInstance();

    public static class ValidationResult {
        public final boolean valid;
        public final String e164Format;   // normalized, storage-safe format
        public final String errorMessage;

        private ValidationResult(boolean valid, String e164Format, String errorMessage) {
            this.valid = valid;
            this.e164Format = e164Format;
            this.errorMessage = errorMessage;
        }

        static ValidationResult ok(String e164) {
            return new ValidationResult(true, e164, null);
        }

        static ValidationResult fail(String message) {
            return new ValidationResult(false, null, message);
        }
    }

    /**
     * Validates and normalizes a phone number for safe storage.
     *
     * @param rawInput    the raw, untrusted user input
     * @param defaultRegion ISO 3166-1 alpha-2 region code (e.g. "US") used
     *                      to interpret numbers without a country code
     */
    public static ValidationResult validate(String rawInput, String defaultRegion) {
        if (rawInput == null || rawInput.isBlank()) {
            return ValidationResult.fail("Phone number is required.");
        }

        // Basic sanitation: cap length and strip characters that have no
        // business being in a phone number, before we even hand it to the parser.
        // This limits the attack surface for malformed/oversized input,
        // though libphonenumber itself is safe against injection-style payloads.
        String sanitized = rawInput.trim();
        if (sanitized.length() > 20) {
            return ValidationResult.fail("Phone number is too long.");
        }
        if (!sanitized.matches("[0-9+()\\-.\\s]+")) {
            return ValidationResult.fail("Phone number contains invalid characters.");
        }

        try {
            PhoneNumber parsed = PHONE_UTIL.parse(sanitized, defaultRegion);

            if (!PHONE_UTIL.isValidNumber(parsed)) {
                return ValidationResult.fail("Phone number is not valid.");
            }

            // Normalize to E.164 (e.g. +14155552671) — the canonical,
            // unambiguous format recommended for storage and downstream use
            // (SMS gateways, dialers, deduplication, etc.)
            String e164 = PHONE_UTIL.format(parsed, PhoneNumberUtil.PhoneNumberFormat.E164);

            return ValidationResult.ok(e164);

        } catch (NumberParseException e) {
            return ValidationResult.fail("Could not parse phone number: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        ValidationResult result = validate("(415) 555-2671", "US");
        if (result.valid) {
            System.out.println("Normalized for storage: " + result.e164Format);
        } else {
            System.out.println("Rejected: " + result.errorMessage);
        }
    }
}